### ISSUE TITLE
ALP correctness audit: solve-rate floor, GRPO normalization, and missing validation

### ISSUE THREAD

**Issue by Alvorecer721** [Alvorecer721/Nemo-RL#36]
## Problem

The current SingleController ALP implementation retains correctness reward and implements a **documented zero-floor variant**. However, its solve-rate floor differs from the original ALP paper, and per-group reward standardization can largely cancel the intended strength of the length penalty. We need to resolve and validate these objective semantics before treating a lower zero-advantage-group rate as better mathematical learning.

This tracks a confirmed objective difference and a validation gap. It does **not** establish that accuracy reward is missing, that the documented variant is an accidental regression, or that ALP reduces held-out accuracy.

Audited main: `6196eabe33686a494268545c54040bbd842098cf`. The relevant reward and estimator function bodies match the frozen source used by the current ALP training arms.

## Current implementation and paper difference

For our binary-correctness, single-turn math recipe, let $c_i\in\{0,1\}$, group size $G=16$, solve rate $p_g=G^{-1}\sum_j c_j$, generated length $L_i$, cap $L_{\max}=12{,}288$, and coefficient $\alpha=0.1$.

Current reward:

$$
r_i=c_i-\alpha p_g\frac{L_i}{L_{\max}}.
$$

The code computes solve rate from explicit binary episode outcomes, using complete prompt-occurrence groups. Its mask counts generated policy tokens and excludes prompt/history/padding. The correctness reward is preserved, and the shaped reward reaches the advantage estimator. See [reward implementation](https://github.com/Alvorecer721/Nemo-RL/blob/6196eabe33686a494268545c54040bbd842098cf/nemo_rl/algorithms/single_controller_utils/rewards.py#L57-L111), [controller wiring](https://github.com/Alvorecer721/Nemo-RL/blob/6196eabe33686a494268545c54040bbd842098cf/nemo_rl/algorithms/single_controller.py#L3338-L3396), and the [explicit zero-floor documentation](https://github.com/Alvorecer721/Nemo-RL/blob/6196eabe33686a494268545c54040bbd842098cf/docs/guides/single-controller.md#L233-L259).

The original ALP paper explicitly floors solve rate at $1/G$, including for unsolved groups. Expressing **that floor change in our existing normalized-length units** gives:

$$
r_i^{\mathrm{floor}}=c_i-\alpha\max(p_g,1/G)\frac{L_i}{L_{\max}}.
$$

The paper's notation for length versus its normalization constant is inconsistent, so this is not a claim that its numeric coefficient transfers unchanged. The floor itself is explicit in [Section 3, Eq. 2 and accompanying explanation](https://arxiv.org/html/2506.05256v2#S3).

With discrete groups, this floor changes only all-wrong groups. Current zero-floor ALP gives those groups no length pressure. A floor can introduce a preference for shorter wrong responses, but adding it alone is not a validated fix.

## Why normalization matters mathematically

Our GRPO recipe uses a group-mean baseline, `normalize_rewards=true`, sample reward standard deviation, and $\epsilon=10^{-6}$. See [GRPO estimator](https://github.com/Alvorecer721/Nemo-RL/blob/6196eabe33686a494268545c54040bbd842098cf/nemo_rl/algorithms/advantage_estimator.py#L73-L114) and [baseline/std computation](https://github.com/Alvorecer721/Nemo-RL/blob/6196eabe33686a494268545c54040bbd842098cf/nemo_rl/algorithms/utils.py#L170-L264).

For a group with identical correctness outcomes and a nonzero length coefficient $k=\alpha\widetilde p_g/L_{\max}$:

$$
r_i=c-kL_i,\qquad r_i-\bar r=-k(L_i-\bar L),\qquad s_r=ks_L.
$$

Consequently:

$$
A_i=\frac{-k(L_i-\bar L)}{ks_L+\epsilon}
\approx-\frac{L_i-\bar L}{s_L}
\quad\text{when }ks_L\gg\epsilon.
$$

The coefficient largely cancels. This already applies to all-correct groups under current ALP, and would also apply to all-wrong groups after adding the floor. Small raw penalties can therefore produce order-one advantages. Actual gradient contribution also depends on token weighting, masks, importance weights, and the loss reduction.

For **16 wrong responses**, eight of length 4,096 and eight of length 12,288:

| Reward/estimator variant | Short-response advantage | Long-response advantage |
|---|---:|---:|
| Current zero floor, group std | 0 | 0 |
| Floor $1/16$, $\alpha=0.1$, group std | +0.9678 | −0.9678 |
| Floor $1/16$, $\alpha=0.001$, group std | +0.9252 | −0.9252 |
| Floor $1/16$, $\alpha=0.1$, group centering only | +0.002083 | −0.002083 |

Reducing $\alpha$ by 100× barely reduces the normalized advantages in this example. Positive advantage for shorter wrong answers is consistent with a length objective; it does not imply a new correctness signal.

If every response is wrong **and has the same capped length**, all rewards are identical. Centering makes every advantage zero, with or without the floor or standardization. Giving every truncated answer the same negative reward cannot create an intra-group preference.

### Minimal reproduction

This PyTorch-only example reproduces the math; it does not require Ray or a GPU:

```python
import torch

lengths = torch.tensor([4096] * 8 + [12288] * 8, dtype=torch.float32)

def advantages(lengths, alpha, floor, normalize):
    rewards = -alpha * floor * lengths / 12288  # all answers wrong
    centered = rewards - rewards.mean()
    std = rewards.std(unbiased=True)
    return centered / (std + 1e-6) if normalize and std > 0 else centered

for name, alpha, floor, normalize in [
    ("current", 0.1, 0.0, True),
    ("floor + std", 0.1, 1 / 16, True),
    ("100x smaller coefficient", 0.001, 1 / 16, True),
    ("floor, centering only", 0.1, 1 / 16, False),
]:
    a = advantages(lengths, alpha, floor, normalize)
    print(name, round(a[0].item(), 6), round(a[-1].item(), 6))

capped = torch.full_like(lengths, 12288)
assert torch.count_nonzero(advantages(capped, 0.1, 1 / 16, True)) == 0
```

Separately, a CPU diagnostic executed unchanged reward/GRPO/helper definitions extracted from the frozen training source with real PyTorch. It reproduced these cases, including all-correct groups and the equal-cap case. This verifies reward/advantage arithmetic, not distributed training or downstream convergence.

## Existing empirical evidence and limits

On a saved initial-SFT DAPO-only probe of **32 prompts × 16 responses**:

- 268/512 responses were truncated; eight groups were all-wrong.
- All eight all-wrong groups had some truncation and varied response lengths; none had all 16 responses truncated.
- Current ALP produced eight zero-advantage groups. Counterfactually adding the floor reduced that count to zero and gave 24 shorter wrong responses positive normalized advantages (threshold $10^{-5}$), without changing any answer's correctness.

These are fixed historical responses, not current training batches or evidence that the floored objective improves training. Local provenance: `session/20260905_114410/dapo_gbs/length_reward_diagnostic.{py,json}` records source/input SHA256 hashes and per-group results; these local files are not public attachments.

The completed **24-update, single-seed** pilot yielded:

| MATH-500 checkpoint | Correct | Mean response tokens | Responses hitting cap |
|---|---:|---:|---:|
| Control | 369/500 (73.8%) | 4,303.9 | 113 |
| Current zero-floor ALP | 366/500 (73.2%) | 4,021.6 | 101 |

ALP used 6.56% fewer tokens. Accuracy difference was −0.6 percentage points, with a paired problem-bootstrap 95% interval of [−4.0, +2.8]. This interval does not capture training-seed variability; asynchronous admitted training groups also differed. The result supports investigating an accuracy/efficiency tradeoff, not a conclusion that ALP is broken or already superior. Local provenance: `session/20260905_114410/overnight_dapo/alp_vs_control.json`.

## Missing validation and proposed closure criteria

- [ ] **Specify the objective.** Decide whether to retain the named zero-floor variant, expose the paper's floor as an explicit option, or adopt it after validation. Document coefficient units and normalization jointly. Preserve historical run/checkpoint identities; do not silently change active runs.
- [ ] **Test reward through advantage estimation.** Existing [ALP tests](https://github.com/Alvorecer721/Nemo-RL/blob/6196eabe33686a494268545c54040bbd842098cf/tests/unit/single_controller/test_alp.py) cover reward arithmetic, group integrity, masks, and configuration guards. Extend coverage for the floor × normalization interaction above: all-wrong/all-correct varied lengths, mixed correctness, equal capped lengths, and $\alpha=0$. Preserve the existing occurrence-group and generated-token invariants.
- [ ] **Measure the kind and strength of signal.** Alongside raw correctness, shaped reward, and penalty, report length-only versus mixed-outcome groups, advantage RMS and loss contribution by group type, joint all-wrong/truncated counts, and short-wrong/malformed-final-answer rates. A falling zero-advantage-group count alone is insufficient.
- [ ] **Run a controlled 70B 2×2 ablation:** floor 0 versus $1/G$, crossed with group std versus group centering only. Use the same starting policy/optimizer state, LR, $\alpha$, sampler, group size, and 12k cap; record admitted samples, effective advantage scale, and gradient norms. Compare at matched updates and compute, with held-out MATH-500 accuracy/tokens and AIME checks. Use multiple training seeds before strong efficacy claims; 92 updates alone is not a convergence criterion.
- [ ] **Judge accuracy and efficiency together.** Report paired accuracy uncertainty, generated-token distributions, truncation, and wall/GPU time. A small accuracy tradeoff may be useful if token savings justify it; improved accuracy and shorter traces are an empirical target, not an acceptance assumption.

For the centering-only arm, `grpo.adv_estimator.normalize_rewards=false` already preserves the group-mean baseline. A batch-normalized alternative would need separate validation: [DLER Section 3.1](https://arxiv.org/html/2510.15110v1#S3) motivates testing group centering followed by batch normalization, but does not establish a 70B benefit here. Our existing [Reinforce++ estimator](https://github.com/Alvorecer721/Nemo-RL/blob/6196eabe33686a494268545c54040bbd842098cf/nemo_rl/algorithms/advantage_estimator.py#L209-L284) uses token-weighted normalization, and [SingleController currently rejects ALP with that estimator](https://github.com/Alvorecer721/Nemo-RL/blob/6196eabe33686a494268545c54040bbd842098cf/nemo_rl/algorithms/single_controller_utils/config.py#L800-L817); it is not an equivalent flag switch.

## Separate, optional truncation/format experiments

ALP is a generated-length penalty, not an explicit `\boxed{}` bonus or a dedicated truncation objective. Adding either is a separate reward-design change. SingleController currently [rejects ALP combined with overlong/stop-properly reward shaping](https://github.com/Alvorecer721/Nemo-RL/blob/6196eabe33686a494268545c54040bbd842098cf/nemo_rl/algorithms/single_controller_utils/rewards.py#L26-L44); overlong **loss filtering** is a distinct mechanism.

A soft overlong penalty near the cap could be tested after explicit integration, but is not a missing requirement of ALP itself. Truncation also does not prove the partial reasoning was wrong: [DAPO Section 3.4](https://arxiv.org/html/2503.14476v2#S3.S4) discusses noise from penalizing otherwise sound unfinished reasoning. Diagnose final-answer extraction, unfinished correct reasoning, and repetitive failure separately before choosing that intervention.
